teachertool/src/state/state.ts

```diff
@@ -15,6 +15,9 @@ export type AppState = {
     validatorPlans: pxt.blocks.ValidatorPlan[] | undefined;
     autorun: boolean;
     confirmationOptions: ConfirmationModalOptions | undefined;
+    blockPickerOptions: BlockPickerOptions | undefined;
```

Comment:
At some point we may want to have a single `modalOptions` field as a compound type, e.g.: `modalOptions: ConfirmationModalOptions | BlockPickerOptions | undefined`

Reply:
I went ahead and combined them, though I took a slightly different approach. May have gone a little overboard with it, but I wanted to avoid blindly casting to the desired types inside the modals themselves, so I ended up making an interface for the options with a modal field we can check against. LMK if you think it's too much.

Reply:
Your changes look fine, but I think it could be simplified a bit:
* Remove `modal` from `AppState`. The presence of a `modalOptions` is enough to know if a modal is showing, and which one.
* Use a union of types for the different modal options, with no base:
```ts
export interface ConfirmationModalOptions {
    modal: "confirmation";
    title: string;
    message: string;
    onCancel: () => void;
    onContinue: () => void;
}

export interface BlockPickerOptions {
    modal: "block-picker";
    criteriaInstanceId: string;
    paramName: string;
}

export type ModalOptions = ConfirmationModalOptions | BlockPickerOptions;
```

* For showing a modal, implement a single `showModal` transform that takes a `ModalOptions` parameter:
```ts
export function showModal(modalOptions: ModalOptions) {
   // dispatch the action
    ...
}
```